### PR TITLE
General parser adf

### LINKED ISSUE #185: ADF Tape21 parser
**Is your feature request related to a problem? Please describe.**
ADF produces the TAPE21 files which are binary files with very specific formatting. Using the dmpkf utility that they have to convert it into an ASCII formatted file we could parse all of the data. The advantage of this is that it is less likely that there would be any changes to the format of the files along with it being universal for parsing things like atom tables which have a different format for nearly every time the atomic positions are listed.

**Describe alternatives you've considered**
Done locally.

### LINKED ISSUE #261: ADF parsers for new version
**Is your feature request related to a problem? Please describe.**
ADF/SCM seems to be making changes to the way their outputs are printing as they have transitioned into their new AMS driver. While there are a lot of similarities between the outputs for the new and old versions there are still changes that are not backwards compatible.

**Describe the solution you'd like**
Implement new parsers for the AMS/ADF driver. Along with this I think it's a good idea to create a parser that can handle the different modules that can be used like NMR and ADF. This would allow us to separate the parsing functions to what is pertinent to each of the modules and make the code more readable. An example of this is the current implementation of the `adf.Output.parse_atom` method as it has multiple if statements if one uses the ADF or NMR modules.

**Additional context**
Not sure if it would be better to have the classes in their own files, such as `nmr_output.py`, `adf_output.py`, and `ams_output.py`. Open to additional ideas.

**Current work**
Currently, I have a working parser that stores the module that we are working with in the `meta` class attribute and we call the methods from the parsing class.

``` python
class Output(six.with_metaclass(OutMeta, Editor)):
    def _parse_data(self, attr):
        try:
            parser = getattr(self.meta['module'], attr)
        except AttributeError as e:
            if str(e).endswith("'"+attr+"'"):
                return
        parser(self)

    def parse_atom(self):
        self._parse_data('parse_atom')

    def __init__(self, file, *args, **kwargs):
        ed = _Editor(file)
        _readf = "A D F"
        _renmr = "N M R"
        _reams = "A M S"
        found = ed.find(_readf, _renmr, _reams, keys_only=True)
        super(Output, self).__init__(file, *args, **kwargs)
        if found[_renmr]:
            self.meta.update({'module': NMR})
        elif found[_readf]:
            self.meta.update({'module': ADF})
        elif found[_reams]:
            self.meta.update({'module': AMS})
        else:
            raise ValueError("Could not identify program from SCM")

```

### PR DESCRIPTION
**Pull request by herbertludowieg**
New implementations:
- Parsing methods for new AMS driver
- Separated the NMR parsing methods from the general ADF class
- Generalized output parsing class for the entire ADF program package
- Parsing methods for an ASCII TAPE21 file

Updates:
- Updated all tests in `test_output.py` to reflect new changes

To do:
- Add tests for the TAPE21 parser

Will close #261 and close #185.

### CODE REVIEW

**Review by tjduigna — COMMENTED** [review]
Hi, thanks for this! The parser looks great, and includes good test coverage. My only concern is the failing tests on ubuntu (and macos but for the same reason). It looks like maybe ipywidgets version drift? Perhaps we can pin it to a last known working version and work out the compatibility issue caused in CI in a later PR. If I remember well the codacy integration was left broken (by me!) so don't worry about that bit.

**Comment by herbertludowieg** [comment]
Thanks! I can look into what's going on with the ipywidgets version. Should this be getting merged directly into master though, or just keep it as is for now?

File: exatomic/adf/tape21.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+        Returns:
+            r_mass (:class:`numpy.ndarray`): Array containing the calculated reduced
+                                             masses.
+        '''
+        cols = ['dx', 'dy', 'dz']
+        # get the isotopic masses of the unique atoms
+        mapper = sym2isomass(symbol)
+        # get a list of the isotopic masses
+        mass = list(map(mapper.get, symbol))
```
Comment (avmarchenko) [inline exatomic/adf/tape21.py:100]: The pattern of `mapper =...; mass = ...` and related lines repeat themselves. I would consider pulling out this functionality into a separate function and converting these static methods into generalized functions elsewhere in the codebase ( in core perhaps?). Not a blocker for me.

File: exatomic/adf/tape21.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        _rekey = r"\bFreq\b"
+        found = self.find(_refreq, _renorm, _recartnorm, keys_only=True)
+        key = self.regex(_rekey, _refreqexc, keys_only=True)
+        # need to do this to ensure that we only match the data in the Freq block
+        found_freq = []
+        for k in key[_rekey]:
+            for f in found[_refreq]:
+                if f-1 == k:
+                    found_freq.append(f)
+                    break
```
Comment (avmarchenko) [inline exatomic/adf/tape21.py:150]: Is this intended to break out of both for loops (as written it breaks out of the innermost only I think)?

File: exatomic/adf/tape21.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+            ndisps = int(self[found[_renorm][0]+1].split()[0])
+            normalmodes = self._dfme(np.array(found[_renorm]), ndisps, idx=0)
+            calc_rmass = True
+        elif found[_recartnorm] and cart:
+            # get the non-mass-weighted normal modes and toss warning
+            ndisps = int(self[found[_recartnorm][0]+1].split()[0])
+            normalmodes = self._dfme(np.array(found[_recartnorm]), ndisps, idx=0)
+            calc_rmass = True
+        else:
+            raise Exception("Something went wrong")
```
Comment (avmarchenko) [inline exatomic/adf/tape21.py:182]: Consider dumping `found` for additional information.

**Review by avmarchenko — APPROVED** [review]
Thank you for this work. Thank you for the unit test coverage. I notice you are merging into a separate branch. I think we follow trunk-based development so I think you'll want to merge into the main branch.

**Pull request closed without merging**